util/resource/resource.go

```diff
@@ -17,61 +17,56 @@ limitations under the License.
 // Package resource implements resource utilites.
 package resource
 
-import "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+import (
+	"sort"
 
-var defaultCreatePriorities = []string{
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var priorityMap = map[string]int{
 	// Namespaces go first because all namespaced resources depend on them.
-	"Namespace",
+	"Namespace": -100,
 	// Custom Resource Definitions come before Custom Resource so that they can be
 	// restored with their corresponding CRD.
-	"CustomResourceDefinition",
+	"CustomResourceDefinition": -99,
 	// Storage Classes are needed to create PVs and PVCs correctly.
-	"StorageClass",
+	"StorageClass": -98,
 	// PVs go before PVCs because PVCs depend on them.
-	"PersistentVolume",
+	"PersistentVolume": -97,
 	// PVCs go before pods or controllers so they can be mounted as volumes.
-	"PersistentVolumeClaim",
+	"PersistentVolumeClaim": -96,
 	// Secrets and ConfigMaps go before pods or controllers so they can be mounted as volumes.
-	"Secret",
-	"ConfigMap",
+	"Secret":    -95,
+	"ConfigMap": -94,
 	// Service accounts go before pods or controllers so pods can use them.
-	"ServiceAccount",
+	"ServiceAccount": -93,
 	// Limit ranges go before pods or controllers so pods can use them.
-	"LimitRange",
+	"LimitRange": -92,
 	// Pods go before ReplicaSets
-	"Pods",
+	"Pods": -91,
 	// ReplicaSets go before Deployments
-	"ReplicaSet",
+	"ReplicaSet": -90,
 	// Endpoints go before Services
-	"Endpoints",
+	"Endpoints": -89,
+}
+
+// PriorityLess returns true if o1 should be created before o2.
+func PriorityLess(o1, o2 unstructured.Unstructured) bool {
+	p1 := priorityMap[o1.GetKind()]
+	p2 := priorityMap[o2.GetKind()]
+	return p1 < p2
 }
 
 // SortForCreate sorts objects by creation priority.
+//
+// Deprecated: use PriorityLess with sort.Slice, sort.SliceStable
+// instead for fast in-place sorting.
 func SortForCreate(resources []unstructured.Unstructured) []unstructured.Unstructured {
-	var ret []unstructured.Unstructured
```

Comment:
If we're deprecating this function we should keep the behaviour the same.

Reply:
And we should update the current usages in CAPI itself to use the new version.

Reply:
> If we're deprecating this function we should keep the behaviour the same.

In what way is it different? Are you referring to the `resources == nil` case? In that case I return a 0-sized slice whereas it currently returns nil, correct?

Reply:
Sorry - I was unclear there. 

I don't think this function should change if this PR deprecates it.

Alternatively it could be retained and the implementation be changed to using PriorityLess with sort.Slice. 

We should choose between deprecating the function or changing the implementation. Doing both in this PR doesn't make sense to me.

Reply:
Oh I see - deprecation shouldn't result in a refactor.

Reply:
Removed the deprecation marker, because its API is much easier to work with than my `sort.Slice` alternative.